crates/pathfinder/src/state/sync/l2.rs

```diff
@@ -93,12 +93,20 @@ pub async fn sync(
             }
         }
 
-        // unwrap is safe as the block hash always exists (unless we query for pending).
+        // Unwrap in both block and state update is safe as the block hash always exists (unless we query for pending).
         let t_update = std::time::Instant::now();
         let state_update = sequencer
             .state_update_by_hash(block.block_hash.unwrap().into())
             .await
             .with_context(|| format!("Fetch state diff for block {:?} from sequencer", next))?;
+        // An extra sanity check for the state update API.
+        assert_eq!(
+            block.block_hash.unwrap(),
+            state_update.block_hash.unwrap(),
+            "State update block hash mismatch, actual {:x}, expected {:x}",
+            block.block_hash.unwrap().0,
+            state_update.block_hash.unwrap().0,
+        );
```

Comment:
I'm not convinced we need this surplus check of the API here.

Reply:
I was thinking at first that there might be a race between the two downloads, but this is a state update by blockhash so it should be safe.

Just wondering about the amount of unwrap on every line here.. But I guess that'd be unrelated refactoring.

Reply:
1. We have an `ensure!` for state root mismatch:
```Rust
anyhow::ensure!(new_root == block.state_root.unwrap(), "State root mismatch");
```
so I think we could have one here too instead of asserting.
2. I don't like to unwrap duplication either..

Reply:
I think that with ensure and deduplication it looks better now.